[image-picker][Android] Add support for selecting multiple media files

bbarthec opened the pull request:
# Why

Part of ENG-2795,
Part of ENG-5617
Part of module audit
Similar to #18102

# How

Used [Intent.EXTRA_ALLOW_MULTIPLE](https://developer.android.com/reference/android/content/Intent#EXTRA_ALLOW_MULTIPLE) flag when launching `ImageLibrary`-based picker.

# Warnings

- When `mediaTypes = 'all'` and `allowsMultipleSelection = true` then all files available on the device are presented - that needs to be investigated further.

# Test Plan

- [x] `ncl`

https://user-images.githubusercontent.com/16623003/177982283-7b5be6b2-fbd4-4689-a554-171fc60baf31.mp4

bbarthec committed:
[image-picker] Add `allowsMultipleSelection` option

bbarthec committed:
[image-picker] Make successful result accept list of assets

bbarthec committed:
[image-picker] Convert to list of assets instead of single asset

bbarthec committed:
[image-picker] Handle multiple selection in `ImageLibraryContract`

bbarthec committed:
[image-picker] Reshape `ImagePickerResponse` to conform to the type from the docs

bbarthec committed:
[image-picker] Adjust `MediaHandler`

bbarthec committed:
[image-picker][ncl] Enable `allowsMultipleSelection` on all platforms (added missing Android support)

bbarthec committed:
[skip ci] Update `CHANGELOG`

barthap commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/contracts/ImageLibraryContract.kt:
```diff
@@ -23,17 +24,48 @@ internal class ImageLibraryContract(
       ImagePickerContractResult.Cancelled()
+    } else if (input.options.allowsMultipleSelection) {
+      val uris = requireNotNull(intent).getClipDataUris()
+      val contentResolver = requireNotNull(appContextProvider.appContext.reactContext) { "React Application Context is null. "}.contentResolver
```
Maybe extract it to a class property:
```kotlin
private val contentResolver: ContentResolver
  get() = requireNotNull(appContextProvider.appContext.reactContext) { 
      "React Application Context is null"
    }.contentResolver
```

Also, remove the dot and space from the end of the error message 😉

barthap commented on packages/expo-image-picker/android/src/main/java/expo/modules/imagepicker/contracts/ImageLibraryContract.kt:
```diff
@@ -23,17 +24,48 @@ internal class ImageLibraryContract(
+    resultSet.add(data)
+  }
+  val clipData = clipData
+  if (clipData == null && resultSet.isEmpty()) {
+    return emptyList()
+  } else if (clipData != null) {
+    for (i in 0 until clipData.itemCount) {
+      val uri = clipData.getItemAt(i).uri
+      if (uri != null) {
+        resultSet.add(uri)
+      }
```
Maybe you'll take some inspiration from:
https://github.com/barthap/expo-music-picker/blob/536168cc1e8ec88371d0dad3aad2833773b82df8/android/src/main/java/expo/community.modules.musicpicker/Utilities.kt#L64-L75
And then you can
```kotlin
for (item in clipData.items) {
 resultSet.add(item.uri)
}

barthap reviewed (commented):
In overall LGTM 💪 I left a few suggestions for tweaks and improvements 😉 

Also, don't forget to run spotless

bbarthec committed:
[image-picker] Create `ClipData.items` extension that conforms to `Iterable<ClipData.Item>`

Create extension function to process stored `ClipData.items` as a `Collection` by implementing `Iterable<ClipData.Item>` interface.

bbarthec committed:
[image-picker] Create `Intent.getAllDataUris` to access all `Uri`s associated with `Intent`

`Intent.getAllDataUris` exposes both `Intent.data` and `Intent.clipData.uri` as a single `List<Uri>`

bbarthec committed:
[image-picker] Adapt `Intent.getAllDataUris` and create `contentResolver` runtime getter in `ImageLibraryContract`
